tracing-flame/src/lib.rs

```diff
@@ -0,0 +1,397 @@
+use std::time::{Duration, Instant};
+use tracing::span::Attributes;
+use tracing::Id;
+use tracing::Subscriber;
+use tracing_subscriber::layer::Context;
+use tracing_subscriber::registry::LookupSpan;
+use tracing_subscriber::registry::SpanRef;
+use tracing_subscriber::Layer;
+
+mod error;
+
+thread_local! {
+    static START: Instant = Instant::now();
+    static LAST_EVENT: Cell<Instant> =  Cell::new(START.with(|f| *f));
+}
+
+/// A Layer that logs span open/close events as folded flame graph stack
+/// samples.
+///
+/// The output of `FlameLayer` emulates the output of commands like `perf` once
+/// they've been collapsed by `inferno-flamegraph`. The output of this layer
+/// should look similar to the output of the following commands:
+///
+/// ```sh
+/// perf record --call-graph dwarf target/release/mybin
+/// perf script | inferno-collapse-perf > stacks.folded
+/// ```
+///
+/// # Sample Counts
+///
+/// Because `tracing-flame` doesn't use sampling, the number at the end of each
+/// folded stack trace does not represent a number of samples of that stack.
+/// Instead, the numbers on each line are the number of nanoseconds since the
+/// last event in the same thread.
+///
+/// # Dropping and Flushing
+///
+/// Because of the way that `tracing` works, if you use a global subscriber the
+/// drop implementations on your various layers will not get called when your
+/// program exits. This means that if you're using a buffered writer as the
+/// inner writer for the `FlameLayer` you're not guaranteed to see all the
+/// events that have been emitted in the file by default.
+///
+/// To ensure all data is flushed when the program exits, `FlameLayer` exposes
+/// the [`flush_on_drop`] function, which returns a [`FlushGuard`]. The `FlushGuard`
+/// will flush the writer when it is dropped. If necessary, it can also be used to manually
+/// flush the writer.
+///
+/// [`flush_on_drop`]: struct.FlameLayer.html#method.flush_on_drop
+/// [`FlushGuard`]: struct.FlushGuard.html
+#[derive(Debug)]
+pub struct FlameLayer<S, W>
+where
+    S: Subscriber,
+    W: Write + 'static,
+{
+    out: Arc<Mutex<W>>,
+    _inner: PhantomData<S>,
+}
+
+/// An RAII guard for managing flushing a global writer that is
+/// otherwise inaccessible.
+///
+/// This type is only needed when using
+/// `tracing::subscriber::set_global_default`, which prevents the drop
+/// implementation of layers from running when the program exits.
+#[must_use]
+#[derive(Debug)]
+pub struct FlushGuard<W>
+where
+    W: Write + 'static,
+{
+    out: Arc<Mutex<W>>,
+}
+
+impl<S, W> FlameLayer<S, W>
+where
+    S: Subscriber + for<'span> LookupSpan<'span>,
+    W: Write + 'static,
+{
+    /// Return a new FlameLayer that outputs all folded stack samples to the
+    /// provided writer
+    pub fn new(writer: W) -> Self {
+        // Initialize the start used by all threads when initializing the
+        // LAST_EVENT when constructing the layer
+        START.with(|_| ());
```

Comment:
I'm not sure if this actually behaves as described in the comment. This will initialize the start time for the _current_ thread --- is it supposed to initialize the same start time for _all_ threads? If it is, this should probably be a `lazy_static` rather than a `thread_local`. I believe that when other threads are spawned, the `START` thread local for those threads will be initialized to the first time _that thread_ accessed it.

Reply:
oh yea you're right, this one has to be global